src/Control/Monad/Catch.hs

```diff
@@ -179,6 +178,21 @@ class MonadCatch m => MonadMask m where
   -- and/or unkillable.
   uninterruptibleMask :: ((forall a. m a -> m a) -> m b) -> m b
 
+  -- | A generalized version of the standard bracket function which
+  -- allows distinguishing different exit cases.
+  --
+  -- @since 0.8.4
+  generalBracket
+    :: m a
+    -- ^ acquire some resource
+    -> (a -> b -> m b)
+    -- ^ cleanup, no exception thrown
+    -> (a -> SomeException -> m ignored)
+    -- ^ cleanup, some exception thrown. The exception will be rethrown
+    -> (a -> m b)
+    -- ^ inner action to perform with the resource
+    -> m b
```

Comment:
I originally tried going that route, but I don't believe any default implementation will be correct for any transformer (unless we used something like `MonadTransControl`).